drivers/mpu9250/MPU9250.cpp

```diff
@@ -349,37 +389,84 @@ void MPU9250::_measure()
 		int16_t		gyro_x;
 		int16_t		gyro_y;
 		int16_t		gyro_z;
-	} report;
+		//uint8_t		ext_data[24];
+	};
 #pragma pack(pop)
 
-	_bulkRead(MPUREG_ACCEL_XOUT_H, (uint8_t *)&report, sizeof(report));
+	// Get FIFO byte count to read and floor it to the report size.
+	int bytes_to_read = get_fifo_count() / sizeof(int_status_report) * sizeof(int_status_report);
 
-	/* TODO: add ifdef for endianness */
-	report.accel_x = swap16(report.accel_x);
-	report.accel_y = swap16(report.accel_y);
-	report.accel_z = swap16(report.accel_z);
-	report.temp = swap16(report.temp);
-	report.gyro_x = swap16(report.gyro_x);
-	report.gyro_y = swap16(report.gyro_y);
-	report.gyro_z = swap16(report.gyro_z);
+	// The max to fit in the read buffer of 512 bytes is:
+	// 13 packets of 38 bytes (including mag) = 494 bytes.
+	const unsigned buf_len = 26*14; //13 * sizeof(int_status_report);
+	uint8_t fifo_read_buf[buf_len];
 
-	m_synchronize.lock();
+	if (bytes_to_read <= 0) {
+		// TODO: increase error counter
+		return;
+	}
 
-	m_sensor_data.accel_m_s2_x = float(report.accel_x) * (MPU9250_ONE_G / 2048.0f);
-	m_sensor_data.accel_m_s2_y = float(report.accel_y) * (MPU9250_ONE_G / 2048.0f);
-	m_sensor_data.accel_m_s2_z = float(report.accel_z) * (MPU9250_ONE_G / 2048.0f);
-	m_sensor_data.temp_c = float(report.temp) / 361.0f + 35.0f;
-	m_sensor_data.gyro_rad_s_x = float(report.gyro_x) * GYRO_RAW_TO_RAD_S;
-	m_sensor_data.gyro_rad_s_y = float(report.gyro_y) * GYRO_RAW_TO_RAD_S;
-	m_sensor_data.gyro_rad_s_z = float(report.gyro_z) * GYRO_RAW_TO_RAD_S;
+	unsigned bytes_done = 0;
+	while (bytes_done < bytes_to_read) {
 
-	m_sensor_data.last_read_time_usec = DriverFramework::offsetTime();
-	m_sensor_data.read_counter++;
+		//DF_LOG_INFO("bytes done: %d, to read: %d", bytes_done, bytes_to_read);
 
-	_publish(m_sensor_data);
+		// XXX DEBUG: always just read one report. To debug ioctl that gets stuck.
+		const unsigned read_len = MIN(bytes_to_read-bytes_done, buf_len);
+		//const unsigned read_len = sizeof(int_status_report);
 
-	m_synchronize.signal();
-	m_synchronize.unlock();
+		memset(fifo_read_buf, 0x0, buf_len);
+		_bulkRead(MPUREG_FIFO_R_W, fifo_read_buf, read_len);
+
+		// Go through buffer, and publish after every report that is complete.
+		// XXX: currently we lose ~ 1/8 packets because _bulkRead is off-by-one which
+		//      we need to discard the last read packet because the last byte is
+		//	unknown.
+		for (unsigned i = 0; i+1+sizeof(int_status_report) < read_len; i += sizeof(int_status_report)) {
```

Comment:
is `i+1+sizeof(int_status_report) < read_len;` correct?

if `i` is being incremented by `sizeof(int_status_report)` each iteration, should it be  `(i+1)*sizeof(int_status_report) < read_len;`?

also, `i` is always accessed as `i+1` - this would appear to discard the first packet rather than the last, as suggested in the comments (unless the fifo is in reverse order).